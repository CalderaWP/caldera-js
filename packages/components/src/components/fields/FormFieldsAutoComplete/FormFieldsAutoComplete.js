import {AutoCompleteField} from '../../fields/AutoCompleteField/AutoCompleteField';

import React from 'react';
import PropTypes from "prop-types";

export const FormFieldsAutoComplete = ({
	label,
	fieldId,
	form,
	description,
	value,
	additionalOptions
								}) => {
	const options = [...additionalOptions];
	const {fields} = form;
	fields.forEach(field => {
		options.push(`%${field.fieldId}%`);
		options.push(field.label);
	});

	return (
		<AutoCompleteField
			label={label}
			fieldId={fieldId}
			description={description}
			value={value}
			options={options}
		/>
	);
};

FormFieldsAutoComplete.IDENTIFIER  = 'fields-autocomplete';


FormFieldsAutoComplete.propTypes = {
	...AutoCompleteField.propTypes,
	additionalOptions: PropTypes.array,
};

FormFieldsAutoComplete.defaultProps = {
	additionalOptions: [],
};
