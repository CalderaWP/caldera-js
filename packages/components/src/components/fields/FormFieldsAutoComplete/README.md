# Form Field Auto-Complete

Auto-complete field, with form field magic tags as the options.

## Examples

### Select from all fields

```jsx
import {} from '@calderajs/components'

<FormFieldsAutoComplete
    label={'Select A Field'}
    onChange={(newValue) => console.log(newValue)}
    fieldId={'selection-field'}
    form={form}
/>
```

### Select from specific field types

```jsx
import {} from '@calderajs/components'

<FormFieldsAutoComplete
    label={'Select A Field'}
    onChange={(newValue) => console.log(newValue)}
    fieldId={'selection-field'}
    form={form}
    allowedTypes={['email', 'text']}
/>
```
