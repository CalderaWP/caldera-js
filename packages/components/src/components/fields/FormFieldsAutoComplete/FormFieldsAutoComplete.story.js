import React, { useState, useMemo } from 'react';
import { storiesOf } from '@storybook/react';
import { FormFieldsAutoComplete } from './FormFieldsAutoComplete';
import { textField, checkboxField } from '../../../factory';

const onChange = newValue => {
	console.log(newValue);
};
const onBlur = event => {};

const STORY_NAME = 'Form Fields Auto Complete field';

function Wrapped({ form }) {
	const [value, onChange] = useState('');
	return (
		<FormFieldsAutoComplete
			label={'Setting Field'}
			onChange={onChange}
			fieldId={'selection-hats'}
			value={value}
			form={form}
		/>
	);
}

const form = {
	id: 'test',
	fields: [checkboxField, textField],
};

storiesOf(STORY_NAME, module).add('Select fields ', () => (
	<Wrapped form={form} />
));
