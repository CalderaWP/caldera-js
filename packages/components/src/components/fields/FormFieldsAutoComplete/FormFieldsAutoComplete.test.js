import * as React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { FormFieldsAutoComplete } from './FormFieldsAutoComplete';

import {
	autoCompleteField,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField,
	radioField
} from '../../../factory';

describe('FormFieldsAutoComplete', () => {
	let onClose;
	let onChange;
	let onBlur;

	const form = {
		fields: [
			selectField,checkboxField
		]
	};
	beforeEach(() => {
		onClose = jest.fn();
		onChange = jest.fn();
		onBlur = jest.fn();
	});

	const {
		label,
		fieldId,
		description,
		value
	} = autoCompleteField;

	it( 'Matches snapshot', () => {
		const component = renderer.create(
			<FormFieldsAutoComplete
				label={label}
				fieldId={fieldId}
				description={description}
				value={value}
				form={form}
			/>
		);

		expect(component.toJSON()).toMatchSnapshot();

	});

	it( 'Additional options merged in', () => {
		const component = renderer.create(
			<FormFieldsAutoComplete
				label={label}
				fieldId={fieldId}
				description={description}
				value={value}
				form={form}
				additionalOptions={['f', 'y']}
			/>
		);

		expect(component.toJSON()).toMatchSnapshot();

	})

});
