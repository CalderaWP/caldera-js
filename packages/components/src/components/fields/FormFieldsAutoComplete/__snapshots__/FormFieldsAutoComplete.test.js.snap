// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormFieldsAutoComplete Additional options merged in 1`] = `
<div
  className="components-base-control"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="autoComplete1"
    >
      Options
    </label>
    <div
      className="components-form-token-field"
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      tabIndex="-1"
    >
      <label
        className="components-form-token-field__label"
        htmlFor="components-form-token-input-1"
      >
        Add item
      </label>
      <div
        className="caldera-field caldera-field-autocomplete components-form-token-field__input-container"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        tabIndex="-1"
      >
        <span
          className="components-form-token-field__token"
          onMouseEnter={undefined}
          onMouseLeave={undefined}
          title={undefined}
        >
          <span
            className="components-form-token-field__token-text"
            id="components-form-token-field__token-text-2"
          >
            <span
              className="screen-reader-text"
            >
              One (1 of 2)
            </span>
            <span
              aria-hidden="true"
            >
              One
            </span>
          </span>
          <button
            aria-describedby="components-form-token-field__token-text-2"
            aria-label="Remove item"
            className="components-button components-icon-button components-form-token-field__remove-token"
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="dashicon dashicons-dismiss"
              focusable="false"
              height={20}
              role="img"
              viewBox="0 0 20 20"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
              />
            </svg>
          </button>
        </span>
        <span
          className="components-form-token-field__token"
          onMouseEnter={undefined}
          onMouseLeave={undefined}
          title={undefined}
        >
          <span
            className="components-form-token-field__token-text"
            id="components-form-token-field__token-text-3"
          >
            <span
              className="screen-reader-text"
            >
              Three (2 of 2)
            </span>
            <span
              aria-hidden="true"
            >
              Three
            </span>
          </span>
          <button
            aria-describedby="components-form-token-field__token-text-3"
            aria-label="Remove item"
            className="components-button components-icon-button components-form-token-field__remove-token"
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="dashicon dashicons-dismiss"
              focusable="false"
              height={20}
              role="img"
              viewBox="0 0 20 20"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
              />
            </svg>
          </button>
        </span>
        <input
          aria-activedescendant={undefined}
          aria-autocomplete="list"
          aria-describedby="components-form-token-suggestions-howto-1"
          aria-expanded={false}
          aria-owns={undefined}
          autoCapitalize={undefined}
          autoComplete={undefined}
          className="components-form-token-field__input"
          disabled={false}
          id="components-form-token-input-1"
          onBlur={[Function]}
          onChange={[Function]}
          role="combobox"
          size={1}
          type="text"
          value=""
        />
      </div>
      <div
        className="screen-reader-text"
        id="components-form-token-suggestions-howto-1"
      >
        Separate with commas
      </div>
    </div>
  </div>
  <p
    className="components-base-control__help"
    id="autoComplete1__help"
  >
    Help text
  </p>
</div>
`;

exports[`FormFieldsAutoComplete Matches snapshot 1`] = `
<div
  className="components-base-control"
>
  <div
    className="components-base-control__field"
  >
    <label
      className="components-base-control__label"
      htmlFor="autoComplete1"
    >
      Options
    </label>
    <div
      className="components-form-token-field"
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyPress={[Function]}
      tabIndex="-1"
    >
      <label
        className="components-form-token-field__label"
        htmlFor="components-form-token-input-0"
      >
        Add item
      </label>
      <div
        className="caldera-field caldera-field-autocomplete components-form-token-field__input-container"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        tabIndex="-1"
      >
        <span
          className="components-form-token-field__token"
          onMouseEnter={undefined}
          onMouseLeave={undefined}
          title={undefined}
        >
          <span
            className="components-form-token-field__token-text"
            id="components-form-token-field__token-text-0"
          >
            <span
              className="screen-reader-text"
            >
              One (1 of 2)
            </span>
            <span
              aria-hidden="true"
            >
              One
            </span>
          </span>
          <button
            aria-describedby="components-form-token-field__token-text-0"
            aria-label="Remove item"
            className="components-button components-icon-button components-form-token-field__remove-token"
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="dashicon dashicons-dismiss"
              focusable="false"
              height={20}
              role="img"
              viewBox="0 0 20 20"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
              />
            </svg>
          </button>
        </span>
        <span
          className="components-form-token-field__token"
          onMouseEnter={undefined}
          onMouseLeave={undefined}
          title={undefined}
        >
          <span
            className="components-form-token-field__token-text"
            id="components-form-token-field__token-text-1"
          >
            <span
              className="screen-reader-text"
            >
              Three (2 of 2)
            </span>
            <span
              aria-hidden="true"
            >
              Three
            </span>
          </span>
          <button
            aria-describedby="components-form-token-field__token-text-1"
            aria-label="Remove item"
            className="components-button components-icon-button components-form-token-field__remove-token"
            disabled={undefined}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            type="button"
          >
            <svg
              aria-hidden="true"
              className="dashicon dashicons-dismiss"
              focusable="false"
              height={20}
              role="img"
              viewBox="0 0 20 20"
              width={20}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 2c4.42 0 8 3.58 8 8s-3.58 8-8 8-8-3.58-8-8 3.58-8 8-8zm5 11l-3-3 3-3-2-2-3 3-3-3-2 2 3 3-3 3 2 2 3-3 3 3z"
              />
            </svg>
          </button>
        </span>
        <input
          aria-activedescendant={undefined}
          aria-autocomplete="list"
          aria-describedby="components-form-token-suggestions-howto-0"
          aria-expanded={false}
          aria-owns={undefined}
          autoCapitalize={undefined}
          autoComplete={undefined}
          className="components-form-token-field__input"
          disabled={false}
          id="components-form-token-input-0"
          onBlur={[Function]}
          onChange={[Function]}
          role="combobox"
          size={1}
          type="text"
          value=""
        />
      </div>
      <div
        className="screen-reader-text"
        id="components-form-token-suggestions-howto-0"
      >
        Separate with commas
      </div>
    </div>
  </div>
  <p
    className="components-base-control__help"
    id="autoComplete1__help"
  >
    Help text
  </p>
</div>
`;
